### PR TITLE
Add structure to Task init

### PR DESCRIPTION
**Pull request by vachillo**
- [x] I have read and agree to the contributing guidelines for [submitting new pull requests](https://github.com/griptape-ai/griptape?tab=readme-ov-file#submitting-pull-requests).

## Describe your changes
- adds `structure` parameter to `Task` init, and updates `add_parent/child` methods to add the `task` to the `structure` if it exists.

## Issue ticket number and link
improves syntax: #896

### CODE REVIEW

File: griptape/tasks/base_task.py [hunk 1]
```diff
@@ -29,26 +29,30 @@ class State(Enum):
...
+            self.structure.add_task(self)
+
     @property
     @abstractmethod
     def input(self) -> BaseArtifact: ...
 
     @property
     def parents(self) -> list[BaseTask]:
-        return [self.structure.find_task(parent_id) for parent_id in self.parent_ids]
+        return [self.structure.find_task(parent_id) for parent_id in self.parent_ids] if self.structure else []
```
Comment (collindutter) [inline griptape/tasks/base_task.py:51]: Explicit `is not None`

Comment (collindutter) [inline griptape/tasks/base_task.py:51]: Use explicit `is not None` check.

File: griptape/tasks/base_task.py [hunk 2]
```diff
@@ -29,26 +29,30 @@ class State(Enum):
...
 
     @property
     def children(self) -> list[BaseTask]:
-        return [self.structure.find_task(child_id) for child_id in self.child_ids]
+        return [self.structure.find_task(child_id) for child_id in self.child_ids] if self.structure else []
```
Comment (collindutter) [inline griptape/tasks/base_task.py:55]: Use explicit `is not None` check.

**Review by collindutter — CHANGES_REQUESTED** [review]
Nice work, can you please update the changelog and show the syntax somewhere in the docs?

**Review by collindutter — DISMISSED** [review]
Looks great! I want to get @dylanholmes eyes on it before merging since he was doing some work around here too.

File: griptape/tasks/base_task.py [hunk 3]
```diff
@@ -29,26 +29,32 @@ class State(Enum):
...
-    structure: Optional[Structure] = field(default=None, init=False)
     context: dict[str, Any] = field(factory=dict, kw_only=True)
     futures_executor_fn: Callable[[], futures.Executor] = field(
         default=Factory(lambda: lambda: futures.ThreadPoolExecutor()),
         kw_only=True,
     )
 
+    def __attrs_post_init__(self) -> None:
+        if self.structure is not None and self not in self.structure.tasks:
+            self.structure.add_task(self)
```
Comment (dylanholmes) [inline griptape/tasks/base_task.py:43]: Could you perform the `self not in self.structure.tasks` check inside `Structure.add_task()` instead of in the client code? (Enforcing the invariant of no duplicate tasks in the `tasks` list is really the responsibility of `Structure`)

File: griptape/tasks/base_task.py [hunk 4]
```diff
@@ -29,26 +29,32 @@ class State(Enum):
...
+
     @property
     @abstractmethod
     def input(self) -> BaseArtifact: ...
 
     @property
     def parents(self) -> list[BaseTask]:
-        return [self.structure.find_task(parent_id) for parent_id in self.parent_ids]
+        return (
+            [self.structure.find_task(parent_id) for parent_id in self.parent_ids] if self.structure is not None else []
```
Comment (dylanholmes) [inline griptape/tasks/base_task.py:52]: I think it might be better to raise an exception if `self.structure` is `None` and this method property is invoked. The task may have parent_ids, but it just can't access them yet because it hasn't been added to the structure. As soon as its added to a structure then `len(task.parents)` will all of a sudden be greater than zero.

File: tests/unit/structures/test_workflow.py [hunk 5]
```diff
@@ -294,28 +294,28 @@ def test_run_topology_1_declarative_mixed(self):
...
-        task2 = PromptTask("test2", id="task2")
-        task3 = PromptTask("test3", id="task3")
-        task4 = PromptTask("test4", id="task4")
+        task2 = PromptTask("test2", id="task2", structure=workflow)
+        task3 = PromptTask("test3", id="task3", structure=workflow)
+        task4 = PromptTask("test4", id="task4", structure=workflow)
         task2.add_parent(task1)
         task3.add_parent("task1")
         task4.add_parents([task2, "task3"])
-        workflow = Workflow(prompt_driver=MockPromptDriver(), tasks=[task1, task2, task3, task4])
```
Comment (dylanholmes) [inline tests/unit/structures/test_workflow.py:304]: The old method should still work, right? Maybe you should add new tests instead of modifying them?

**Review by dylanholmes — COMMENTED** [review]
Nice work!

Just a couple of minor comments

File: CHANGELOG.md [hunk 6]
```diff
@@ -9,6 +9,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 - `AstraDbVectorStoreDriver` to support DataStax Astra DB as a vector store.
 - Ability to set custom schema properties on Tool Activities via `extra_schema_properties`.
+- Parameter `structure` to `BaseTask`. `BaseTask.add_parent/child` will now call `self.structure.add_task` if possible.
```
Comment (collindutter) [inline CHANGELOG.md:12]: > `BaseTask.add_parent/child` will now call `self.structure.add_task` if possible.

This should probably go under `Changed`

File: griptape/structures/structure.py [hunk 7]
```diff
@@ -227,7 +227,11 @@ def context(self, task: BaseTask) -> dict[str, Any]:
         return {"args": self.execution_args, "structure": self}
 
     def resolve_relationships(self) -> None:
-        task_by_id = {task.id: task for task in self.tasks}
+        task_by_id = {}
+        for task in self.tasks:
+            if task.id in task_by_id:
+                raise ValueError(f"Task with id {task.id} already exists.")
```
[the remark this answers is not in the document]
Comment (collindutter) [reply griptape/structures/structure.py:238]: @vachillo is there a reason for not resolving this one?

File: tests/unit/structures/test_workflow.py [hunk 8]
```diff
@@ -781,8 +809,6 @@ def _validate_topology_1(workflow) -> None:
         assert len(workflow.tasks) == 4
         assert workflow.input_task.id == "task1"
         assert workflow.output_task.id == "task4"
-        assert workflow.input_task.id == workflow.tasks[0].id
-        assert workflow.output_task.id == workflow.tasks[-1].id
```
Comment (collindutter) [inline tests/unit/structures/test_workflow.py:785]: Why did you remove these?

File: griptape/structures/structure.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (vachillo) [reply griptape/structures/structure.py:238]: this is double-checking that no duplicate tasks are in `self.tasks`, not trying to find the task.

File: tests/unit/structures/test_workflow.py (hunk 8 above)
Comment (vachillo) [reply tests/unit/structures/test_workflow.py:785]: because this is an incorrect assumption. `input_task` is the first task in the `ordered_tasks` list, not `tasks`. same for `output_task` except the last task.

Comment (collindutter) [reply tests/unit/structures/test_workflow.py:785]: Can you update the test to use `ordered_tasks` then?

File: griptape/structures/structure.py [hunk 9]
```diff
@@ -227,7 +232,11 @@ def context(self, task: BaseTask) -> dict[str, Any]:
         return {"args": self.execution_args, "structure": self}
 
     def resolve_relationships(self) -> None:
-        task_by_id = {task.id: task for task in self.tasks}
+        task_by_id = {}
+        for task in self.tasks:
+            if task.id in task_by_id:
+                raise ValueError(f"Duplicate task with id {task.id} found.")
```
Comment (dylanholmes) [inline griptape/structures/structure.py:238]: Much better error message!

File: griptape/tasks/base_task.py [hunk 10]
```diff
@@ -76,21 +84,35 @@ def add_parents(self, parents: list[str | BaseTask]) -> None:
...
         parent_id = parent if isinstance(parent, str) else parent.id
 
         if parent_id not in self.parent_ids:
             self.parent_ids.append(parent_id)
 
+        if parent_task is not None:
+            parent_task.add_child(self.id)
+
+            if self.structure is not None:
+                self.structure.add_task(parent_task)
```
Comment (dylanholmes) [inline griptape/tasks/base_task.py:97]: Maybe a little simpler without the `parent_task` temporary variable:

```py
if isinstance(parent, BaseTask):
    parent.add_child(self.id)

    if self.structure is not None:
        self.structure.add_task(parent)
```

**Review by dylanholmes — DISMISSED** [review]
Sweet!

Just one optional suggestion

**Review by collindutter — APPROVED** [review]
Nice 😎
